filestream/filestream.go

```diff
@@ -0,0 +1,158 @@
+	}
+
+	err = watcher.Add(file_name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add file '%s' to the watcher: %w", file_name, err)
+	}
+
+	return &FileStream{
+		file_name:   file_name,
+		tailEnabled: tail,
+		reader:      file,
+		watcher:     watcher,
+		ctx:         ctx,
+		done:        make(chan bool),
+		logComplete: make(chan bool),
+	}, nil
+}
+
+//-------------------------------------------------------------------------------------------------
+func (s *FileStream) TailEnabled() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.tailEnabled
+}
+
+func (s *FileStream) DisableTail() {
+	if !s.TailEnabled() {
+		return
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.tailEnabled = false
+	close(s.logComplete) // Tell the reader to stop waiting for more content
+}
+
+//-------------------------------------------------------------------------------------------------
+// Closes the fsnotify watcher and the underlying file stream
+func (s *FileStream) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Prevent double-close
+	if s.isClosed {
+		return nil
+	}
+	s.isClosed = true
+
+	// Send a "quit" message to the reader goroutine
+	close(s.done)
+
+	// Close both the watcher and the file reader,
+	// Return possible errors from the file reader since the watcher always returns nil
+	s.watcher.Close()
+	return s.reader.Close()
+}
+
+//-------------------------------------------------------------------------------------------------
+// Blocks until there is a write event on the file or we're asked to stop waiting and/or reading
+func (s *FileStream) waitForChanges() {
+	for {
+		select {
+		// Some write occurred and we should see if we could read something
+		case event := <-s.watcher.Events:
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				return
+			}
+		// We were asked to stop tailing the file
+		case <-s.logComplete:
+			return
+		// We were asked to stop reading the file
+		case <-s.done:
+			return
+		// The context has been cancelled and we should stop
+		case <-s.ctx.Done():
+			return
+		}
+	}
+}
+
+//-------------------------------------------------------------------------------------------------
+// Returns true if we are done reading the stream or were asked to stop
+func (s *FileStream) shouldStop() bool {
+	select {
+	case <-s.done:
+		return true
+	case <-s.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+//-------------------------------------------------------------------------------------------------
+// Implements the io.Reader interface
+// Returns the data available in the stream and blocks for more data if the stream is empty.
+func (s *FileStream) Read(buffer []byte) (int, error) {
+	for {
+		if s.shouldStop() {
+			return 0, io.EOF
+		}
+		read_bytes, err := s.reader.Read(buffer)
+
+		// How we handle the EOF depends on the current tail mode
+		if err == io.EOF {
+			if s.TailEnabled() {
+				err = nil
+			} else {
+				s.Close()
```

Comment:
I think you probably want to propagate the `io.EOF` here and let the user call `Close` themselves. That behavior would be more consistent with other `io.Reader` implementations.